src/util/codec/mysql/decimal.rs

```diff
@@ -11,770 +11,2008 @@
+    if encoded.len() < 2 {
+        return Err(box_err!("decimal too short: {} < 2", encoded.len()));
     }
 
-    /// Construct a decimal using the specified coefficient, exponent and fsp.
-    ///
-    /// If constucted decimal contains more fraction digit than fsp, they will be
-    /// rounded.
-    pub fn new(coeff: BigInt, mut exp: i32, mut fsp: u8) -> Decimal {
-        fsp = cmp::min(fsp, MAX_FSP);
+    let precision = encoded[0];
+    let frac_cnt = encoded[1];
+    if precision < frac_cnt {
+        return Err(box_err!("invalid decimal, precision {} < frac_cnt {}",
+                            precision,
+                            frac_cnt));
+    }
+    let int_cnt = precision - frac_cnt;
+    let int_word_cnt = int_cnt / DIGITS_PER_WORD;
+    let frac_word_cnt = frac_cnt / DIGITS_PER_WORD;
+    let int_left = (int_cnt - int_word_cnt * DIGITS_PER_WORD) as usize;
+    let frac_left = (frac_cnt - frac_word_cnt * DIGITS_PER_WORD) as usize;
+    let int_len = (int_word_cnt * WORD_SIZE + DIG_2_BYTES[int_left]) as usize;
+    let frac_len = (frac_word_cnt * WORD_SIZE + DIG_2_BYTES[frac_left]) as usize;
+    Ok(int_len + frac_len + 2)
+}
+
+/// `count_leading_zeroes` returns the number of leading zeroes that can be removed from int.
+fn count_leading_zeroes(i: u8, word: u32) -> u8 {
+    let (mut c, mut i) = (0, i as usize);
+    while TEN_POW[i] > word {
+        i -= 1;
+        c += 1;
+    }
+    c
+}
 
-        if coeff.is_zero() {
-            exp = 0;
+/// `count_trailing_zeroes` returns the number of trailing zeroes that can be removed from fraction.
+fn count_trailing_zeroes(i: u8, word: u32) -> u8 {
+    let (mut c, mut i) = (0, i as usize);
+    while word % TEN_POW[i] == 0 {
+        i += 1;
+        c += 1;
+    }
+    c
+}
+
+/// `add` adds a and b and carry, stores the sum and new carry.
+fn add(a: u32, b: u32, carry: &mut u32, res: &mut u32) {
+    let sum = a + b + *carry;
+    if sum >= WORD_BASE {
+        *res = sum - WORD_BASE;
+        *carry = 1;
+    } else {
+        *res = sum;
+        *carry = 0;
+    }
+}
+
+/// `fix_word_cnt_err` limits word count in `word_buf_len`.
+fn fix_word_cnt_err(int_word_cnt: u8, frac_word_cnt: u8, word_buf_len: u8) -> Res<(u8, u8)> {
+    if int_word_cnt + frac_word_cnt > word_buf_len {
+        if int_word_cnt > word_buf_len {
+            return Res::Overflow((word_buf_len, 0));
         }
+        return Res::Truncated((int_word_cnt, word_buf_len - int_word_cnt));
+    }
+    Res::Ok((int_word_cnt, frac_word_cnt))
+}
 
-        let d = Decimal {
-            coeff: coeff,
-            exp: exp,
-            fsp: fsp,
-        };
+/// `sub` substracts rhs and carry from lhs, store the diff and new carry.
+fn sub(lhs: u32, rhs: u32, carry: &mut i32, res: &mut u32) {
+    let diff = lhs as i32 - rhs as i32 - *carry;
+    if diff < 0 {
+        *carry = 1;
+        *res = (diff + WORD_BASE as i32) as u32;
+    } else {
+        *carry = 0;
+        *res = diff as u32;
+    }
+}
```

Comment:
Need to port sub2()?
https://github.com/pingcap/tidb/blob/master/mysql/mydecimal.go#L89